### PR TITLE
Treat concurrent cosmos_cache Variable insert IntegrityError as benign

### LINKED ISSUE #2723: [Bug] Concurrent DAG parsing can raise IntegrityError on cosmos_cache__* Airflow Variable insert
### Astronomer Cosmos Version

1.13.1

### LoadMode

DBT_LS

### Deployment

Astronomer

### Deployment details

Internal report originating from Zendesk ticket [91761](https://astronomer.zendesk.com/agent/tickets/91761). Multiple Cosmos DAGs deployed on a single Airflow environment; metadata database is Postgres.

### What happened?

When two scheduler or DAG-processor processes parse a Cosmos DAG concurrently and neither finds an existing `cosmos_cache__*` Airflow Variable, both compute the dbt ls cache value and both attempt to write it via `Variable.set`. The unique constraint on `variable.key` lets one INSERT succeed and causes the other to raise a SQLAlchemy `IntegrityError` (Postgres `UniqueViolation`) like:

```
[SQL: INSERT INTO variable (key, val, description, is_encrypted)
 VALUES (%(key)s, %(val)s, %(description)s, %(is_encrypted)s) RETURNING variable.id]
[parameters: {'key': 'cosmos_cache__<dag_id>', ...}]
```

The error is benign in that subsequent parses find the existing row and the read path is used, so the symptom self-heals after the first occurrence. It is, however, noisy in scheduler/DAG-processor logs and creates the impression of a real failure.

`cosmos/dbt/graph.py:_save_cache_to_variable` already wraps `Variable.set` in a `try/except AirflowRuntimeError`, but does not catch SQLAlchemy `IntegrityError` (or the Airflow 3 SDK equivalent) raised by the unique-constraint violation. The fix is to treat the duplicate-key race as benign (the value we wanted to write is already there) and log it at debug level rather than letting it propagate.

**Expected behaviour:** Concurrent parsers racing on the same `cosmos_cache__*` key should both complete without raising; whichever process wins the INSERT writes the value, the other treats the duplicate-key error as a no-op.

### Relevant log output

```shell
sqlalchemy.exc.IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "variable_key_key"
DETAIL:  Key (key)=(cosmos_cache__<redacted>) already exists.
[SQL: INSERT INTO variable (key, val, description, is_encrypted) VALUES (%(key)s, %(val)s, %(description)s, %(is_encrypted)s) RETURNING variable.id]
[parameters: {'key': 'cosmos_cache__<redacted>', ...}]
```

### How to reproduce

1. Deploy two or more Cosmos DAGs using `LoadMode.DBT_LS` with caching enabled (default).
2. Ensure no `cosmos_cache__<dag_id>` Airflow Variable exists yet for the target DAG.
3. Trigger concurrent DAG parses on the same DAG (for example, start a scheduler and a DAG processor at the same time, force a rapid restart, or scale out the DAG processor).
4. Observe one of the parsers raises the `IntegrityError` above while the other completes successfully and writes the Variable.

A more deterministic local repro: instrument `_save_cache_to_variable` with a `time.sleep` after the `Variable.get` lookup but before `Variable.set`, then run two DAG-parse invocations against the same metadata DB simultaneously.

### Anything else?

- Symptom is one-shot per cache key: subsequent runs read the existing row and succeed.
- Two workarounds exist (set `AIRFLOW__COSMOS__REMOTE_CACHE_DIR` to use object storage, or set `AIRFLOW__COSMOS__ENABLE_CACHE_DBT_LS=0` to disable the cache), but both have tradeoffs (remote cache adds parse latency; disabling the cache increases parse time on every cycle). Neither should be required to avoid the noisy traceback.
- Same race likely applies to the YAML selectors cache (`enable_cache_dbt_yaml_selectors`) since it shares the `_save_cache_to_variable` write path.

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by pankajkoti**
## Summary

Concurrent DAG parses that both miss the `cosmos_cache__*` Airflow Variable can race on the INSERT. The unique constraint on `variable.key` lets one writer win and makes the other raise a SQLAlchemy `IntegrityError` (Postgres `UniqueViolation`), which propagated as a noisy traceback in scheduler/DAG-processor logs even though the cache value was already written.

`_save_cache_to_variable` now treats this duplicate-key race as a benign no-op: the loser logs it at debug level instead of raising. This is handled on both the Airflow 2 path (direct metadata-DB write) and the Airflow 3 path (alongside the existing `AirflowRuntimeError` handling), so it also covers the YAML selectors cache, which shares the same write path.

- Add `_log_concurrent_cache_write` helper for the shared debug log.
- Catch `IntegrityError` in both branches of `_save_cache_to_variable`.
- Add tests asserting the duplicate-key error is swallowed with a debug log for both the dbt ls and YAML selectors caches.

## Related Issue(s)

closes #2723
closes BOSS-416

## Breaking Change?

No.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cosmos/dbt/graph.py [hunk 1]
```diff
@@ -629,17 +629,38 @@ def get_dbt_yaml_selectors_cache_key_args(self, impl_version: str) -> list[str]:
...
+        from sqlalchemy.exc import IntegrityError
+
         try:
             from airflow.sdk.exceptions import AirflowRuntimeError
         except ImportError:
-            Variable.set(self.cache_key, cache_dict, serialize_json=True)
+            try:
+                Variable.set(self.cache_key, cache_dict, serialize_json=True)
+            except IntegrityError:
+                self._log_concurrent_cache_write(cache_name)
```
[the remark this answers is not in the document]
Comment (pankajkoti) [reply cosmos/dbt/graph.py:649]: Narrowing the catch would mean matching on the unique-violation signal per backend (Postgres SQLSTATE 23505, MySQL errno 1062, SQLite "UNIQUE constraint failed"), since the metadata DB is not always Postgres. That adds DB-specific, message-fragile code to guard against. 

The write is also self-healing: if an unexpected integrity error occurs, the value is recomputed and rewritten on the next DAG parse, so nothing is permanently lost.

**Review by tatiana — APPROVED** [review]
Thanks for fixing this, @pankajkoti ! It is a good solution.
